src/lib/structures/profiling.ml

```diff
@@ -196,44 +226,47 @@ let new_instance_of axiom inst loc kept =
     else
       {ii with ignored = ii.ignored + 1}
   in
-  state.instances_map := MS.add axiom ii !(state.instances_map)
+  state.instances_map <- MS.add axiom ii state.instances_map
 
 let conflicting_instance axiom loc =
   let ii =
-    try MS.find axiom !(state.instances_map)
+    try MS.find axiom state.instances_map
     with Not_found -> empty_inst_info loc
   in
   let ii = {ii with confl = ii.confl + 1} in
   assert (ii.loc == loc);
-  state.instances_map := MS.add axiom ii !(state.instances_map)
+  state.instances_map <- MS.add axiom ii state.instances_map
 
 let decision_on_instance axiom_name =
   try
     let ii =
-      MS.find axiom_name !(state.instances_map)
+      MS.find axiom_name state.instances_map
     in
     let ii = {ii with decided = ii.decided + 1} in
     (*assert (ii.loc == loc);*)
-    state.instances_map := MS.add axiom_name ii !(state.instances_map)
+    state.instances_map <- MS.add axiom_name ii state.instances_map
   with Not_found -> ()
 
 
 let decision d origin =
-  incr state.decisions;
-  incr state.decision_lvl;
-  if not (!(state.decision_lvl) = d) then begin
+  state.decisions <- state.decisions + 1;
+  state.decision_lvl <- state.decision_lvl + 1;
+  if state.decision_lvl <> d then begin
     Printer.print_err
       "state.decision_lvl = %d et d = %d"
-      !(state.decision_lvl) d;
+      state.decision_lvl d;
+    (* TODO(Steven): if nobody knows how to use the previous print to debug
+       anything, I suggest we either remove `d` or get rid the invariant:
+       #(calls to Profiling.decision) = d. *)
```

Comment:
Why not just register the decision level of the SAT instead of incrementing it? This invariant forces the Profiling module to be used in a very specific manner, which is undocumented.

Reply:
When we make a decision the decision level is always incremented (by definition). As I say, it just looks like a safeguard. It should be correct to call `reset_dlevel d` here instead of incrementing `state.decision_lvl`, I'm not sure having the decision level in the profiling matters much anyways.

Reply:
That's why my comment mentions that, if we don't know if there is a deep meaning behind the error message and what it is, maybe we should just get rid of it. If we ever get in that case and get this error message, I'm not sure we know how to interpret it.

Reply:
This error message means that the SAT solver is not calling the profiling module in the expected way, and that the profiling module may be accumulating incorrect statistics, because the state that the profiling module expects the solver to be in and the state the solver actually is in have diverged (they have different decision levels). I am trying to reply to your comment and explain what the error message means.